src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java

```diff
@@ -77,6 +91,37 @@ public void rebuildSessionFromFrame(MenuSession menuSession) throws CommCareSess
                         break;
                     }
                 }
+            } else if (screen instanceof QueryScreen) {
+                QueryScreen queryScreen = (QueryScreen)screen;
+                RemoteQueryDatum neededDatum = (RemoteQueryDatum) menuSession.getSessionWrapper().getNeededDatum();
+                boolean done = false;
+                for (StackFrameStep step : steps) {
+                    if (step.getId().equals(neededDatum.getDataId())) {
+                        URI uri = null;
+                        try {
+                            uri = new URI(step.getValue());
+                        } catch (URISyntaxException e) {
+                            e.printStackTrace();
+                            throw new CommCareSessionException("Query URL format error: " + e.getMessage());
+                        }
+                        try {
+                            ExternalDataInstance searchDataInstance = caseSearchHelper.getRemoteDataInstance(
+                                queryScreen.getQueryDatum().getDataId(),
+                                queryScreen.getQueryDatum().useCaseTemplate(),
+                                uri
+                            );
+                            queryScreen.setQueryDatum(searchDataInstance);
+                            screen = menuSession.getNextScreen(false);
+                            done = true;
+                        } catch (InvalidStructureException | IOException | XmlPullParserException | UnfullfilledRequirementsException e) {
+                            e.printStackTrace();
+                            throw new CommCareSessionException("Query response format error: " + e.getMessage());
```

Comment:
Apologies if this pattern is in use elsewhere and you were just copying, but it would be best for this exception to be initialized with `e` as a source so that any reported errors actually carry the full stack trace.